ipa-pwd-extop: don't check password policy for non-Kerberos account set by DM or a passsync manager

abbra opened the pull request:
Password changes performed by cn=Directory Manager are excluded from
password policy checks according to [1]. This is correctly handled by
ipa-pwd-extop in case of a normal Kerberos principal in IPA. However,
non-kerberos accounts were not excluded from the check.

As result, password updates for PKI CA admin account in o=ipaca were
failing if a password policy does not allow a password reuse. We are
re-setting the password for PKI CA admin in ipa-replica-prepare in case
the original directory manager's password was updated since creation of
`cacert.p12`.

Do password policy check for non-Kerberos accounts only if it was set by
a regular user or admin. Changes performed by a cn=Directory Manager and
passsync managers should be excluded from the policy check.

Fixes: https://pagure.io/freeipa/issue/7181
Signed-off-by: Alexander Bokovoy <abokovoy@redhat.com>

[1] https://access.redhat.com/documentation/en-us/red_hat_directory_server/10/html/administration_guide/user_account_management-managing_the_password_policy

abbra commented:
Note that to test this according to the ticket 7181, we need to prepare a second replica so that password reuse for PKI CA admin becomes a reality for a password policy check. Instead, it might be easier to just try to change a password of sudo sysaccount. Thus, patch would be a WIP until a test could be added.

abbra commented:
Added tests.

rcritten commented on ipatests/test_integration/test_commands.py:
```diff
@@ -134,6 +134,84 @@ def test_change_sysaccount_password_issue7561(self):
+                                           use_dirman=False)
+        tasks.ldappasswd_sysaccount_change(sysuser, new_passwd,
+                                           new_passwd, master,
+                                           use_dirman=True)
```
I don't know yet if it is the test or something else but I added in another call to tasks.ldappasswd_sysaccount_change() with use_dirman=False and the test passed when it should have failed.

abbra replied:
We have three different paths, original patch only covered pre_mod. I added change password op and pre_add. We now always check policy but for dirman case we ignore returned constraint violation.

simo5 replied:
Don, non-krb users even have a password history ?

rcritten commented:
Subsequent password changes are still successful. Given that it should fail I added this to the test;

diff --git a/ipatests/test_integration/test_commands.py b/ipatests/test_integ
ration/test_commands.py
index 8fdb6b891..7475c3634 100644
--- a/ipatests/test_integration/test_commands.py
+++ b/ipatests/test_integration/test_commands.py
@@ -211,6 +211,9 @@ class TestIPACommand(IntegrationTest):
         tasks.ldappasswd_sysaccount_change(sysuser, new_passwd,
                                            new_passwd, master,
                                            use_dirman=True)
+        tasks.ldappasswd_sysaccount_change(sysuser, new_passwd,
+                                           original_passwd, master,
+                                           use_dirman=False)

The password change is successful when it should raise an error about re-use AFAICT.

abbra commented:
With latest changes we properly process password changes for non-Kerberos users -- I verified it in manual tests. However, the test does not fail for an additional password change somehow. I'll look at that next week.

1. Password change as a user
```
DEBUG - start_tls - Received extended operation request with OID 1.3.6.1.4.1.1466.20037
DEBUG - start_tls - Start TLS extended operation request confirmed.
DEBUG - ipa-pwd-extop - Attempting OTP authentication for 'uid=forcedpolicy,cn=sysaccounts,cn=etc,$SUFFIX'.
DEBUG - PBKDF2_SHA256 - Comparing password
DEBUG - ipa-pwd-extop - no krbPrincipalName in user entry: uid=forcedpolicy,cn=sysaccounts,cn=etc,$SUFFIX
ipa-lockout-plugin - preop returning 0: success
PBKDF2_SHA256 - Comparing password
DEBUG - backend_passwdmod_extop - extopdn = uid=forcedpolicy,cn=sysaccounts,cn=etc,$SUFFIX
DEBUG - ipa-pwd-extop - Received extended operation request with OID 1.3.6.1.4.1.4203.1.11.1
DEBUG - PBKDF2_SHA256 - Comparing password
DEBUG - PBKDF2_SHA256 - Generated hash {PBKDF2_SHA256}some hash
DEBUG - ipa-pwd-extop - Not a kerberos user, ignore krb attributes
DEBUG - ipa-pwd-extop - success
```

2. Password change as `cn=Directory Manager`:
```
DEBUG - backend_passwdmod_extop - extopdn = uid=forcedpolicy,cn=sysaccounts,cn=etc,$SUFFIX
DEBUG - ipa-pwd-extop - Received extended operation request with OID 1.3.6.1.4.1.4203.1.11.1
DEBUG - PBKDF2_SHA256 - Generated hash {PBKDF2_SHA256}some hash
DEBUG - ipa-pwd-extop - Not a kerberos user, ignore krb attributes
DEBUG - ipa-pwd-extop - success
```

3. New password change as a user using old password, leads to rejection:
```
DEBUG - start_tls - Received extended operation request with OID 1.3.6.1.4.1.1466.20037
DEBUG - start_tls - Start TLS extended operation request confirmed.
DEBUG - ipa-pwd-extop - Attempting OTP authentication for 'uid=forcedpolicy,cn=sysaccounts,cn=etc,$SUFFIX'.
DEBUG - PBKDF2_SHA256 - Comparing password
DEBUG - ipa-pwd-extop - no krbPrincipalName in user entry: uid=forcedpolicy,cn=sysaccounts,cn=etc,$SUFFIX
DEBUG - PBKDF2_SHA256 - Comparing password
DEBUG - backend_passwdmod_extop - extopdn = uid=forcedpolicy,cn=sysaccounts,cn=etc,$SUFFIX
DEBUG - ipa-pwd-extop - Received extended operation request with OID 1.3.6.1.4.1.4203.1.11.1
DEBUG - PBKDF2_SHA256 - Comparing password
DEBUG - ipa-pwd-extop - Password reuse not permitted
```

tiran commented on daemons/ipa-slapi-plugins/ipa-pwd-extop/common.c:
```diff
@@ -465,6 +465,23 @@ int ipapwd_entry_checks(Slapi_PBlock *pb, struct slapi_entry *e,
+     * Note that slapi_entry_attr_find() returns 0 if attr exists, -1 for absence
+     */
+    if (*is_krb) {
+	Slapi_Attr *attr_prname = NULL;
```
It looks like your patch uses TAB while the rest of the code uses 4 space indention.

tiran commented on ipatests/test_integration/test_commands.py:
```diff
@@ -134,6 +135,95 @@ def test_change_sysaccount_password_issue7561(self):
+
+        # Add a system account and add it to a group managed by the policy
+        base_dn = str(master.domain.basedn)  # pylint: disable=no-member
+        tf = NamedTemporaryFile()
```
I have seen this hack a couple of times and still don't like it. It creates a local temporary file and assumes, that it's also a valid name on a remote machine. How about a new helper function in ``ipatests.pytest_plugins.integration`` to properly create a temporary file on the remote host?

```
def upload_temp_contents(host, contents, encoding='utf-8'):
    """Upload contents to a temporary file

    :param host: Remote host instance
    :param contents: file content (str, bytes)
    :param encoding: file encoding
    :return: Temporary file name
    """
    result = host.run_command(['mktemp'])
    tmpname = result.stdout_text.strip()
    host.put_file_contents(tmpname, contents, encoding=encoding)
    return tmpname
```

abbra commented:
@tiran I have pushed an update that addresses your concerns. I still need to find out what's wrong with a test but I changed an assertion for a succeeding `ldappasswd` to ensure test fails if the last password change succeeds.

tiran commented on ipatests/test_integration/test_commands.py:
```diff
@@ -134,6 +134,94 @@ def test_change_sysaccount_password_issue7561(self):
+            if e.returncode != 1:
+                raise
+        else:
+            assert "Password change violating policy did not fail" is False
```
That's an unusual way to raise an assertion. ``"message" is False`` is evaluated first. I prefer ``pytest.fail("message")``.

abbra commented:
@tiran changed assert to fail(message).

abbra commented:
Note that the test still fails, so this PR has to stay yet.

rcritten commented:
@abbra Did you want to do any more work on this, it LGTM?

abbra commented:
@simo5, could you please look at this patchset? I'm puzzled why it doesn't do what it should -- namely, the test for intentional password reuse passes while it should fail.

abbra commented on daemons/ipa-slapi-plugins/ipa-pwd-extop/ipa_pwd_extop.c:
```diff
@@ -561,16 +561,21 @@ static int ipapwd_chpwop(Slapi_PBlock *pb, struct ipapwd_krbcfg *krbcfg)
 			rc = LDAP_CONSTRAINT_VIOLATION;
+			goto free_and_return;
 		}
-		goto free_and_return;
```
This is what I did in the first variant of this patch but then I realized the point of ipapwd_CheckPolicy() is also to update the password history. If we are applying a password history for non-Kerberos accounts, we should call ipapwd_CheckPolicy() so that ipapwd_check_policy() would update the history.

simo5 replied:
Yeah then it's time we split Checks from Updates, it's hard to reason about an interface when it starts doing things it does not declare.

simo5 commented on daemons/ipa-slapi-plugins/ipa-pwd-extop/common.c:
```diff
@@ -465,6 +465,23 @@ int ipapwd_entry_checks(Slapi_PBlock *pb, struct slapi_entry *e,
+        int has_prname = slapi_entry_attr_find(e, "krbPrincipalName", &attr_prname);
+        int has_memberOf = slapi_entry_attr_find(e, "memberOf", &attr_memberof);
+        if ((has_prname == -1) && (has_memberOf == 0)) {
+            *is_krb = 0;
```
Ok I think this is the cause of password history being simply not set therefore never matching.
As this line here:
  https://github.com/freeipa/freeipa/blob/0aeccc08524931b0e37773b4836734bc32eca7f4/daemons/ipa-slapi-plugins/ipa-pwd-extop/prepost.c#L1059
causes this code to bail out before ipapwd_setPasswordHistory() is called a few lines below.

tiran commented:
@abbra I rebased your PR and addressed all merge conflicts.

abbra commented:
I'm at the point where the code works locally but fails in PR CI. I'm going to redo my local test environment to make sure I'm replicating the same behavior as in PR CI.

rcritten commented:
rebased for temp commit

abbra commented:
Rebased to master, let's start from scratch with tests.

rcritten commented:
The test is failing because the forcedpolicy sysaccount user has no passwordHistory.

It has no history because policy isn't being applied, I'm assuming because the sysaccount user doesn't have a krbPwdPolicyReference set. It may be an issue with the class of service configuration.

rcritten commented:
I think the issue is that this is reported against a user in o=ipaca and using global policy and the test is using a sysaccount and creating its own policy. I modified it locally and it passed, need to do some more research on why.

abbra commented:
It should not matter -- global policy is applied when there is no any other referenced policy. The issue I saw was error on processing missing reference (because the object has no objectclass that allows the policy inferring).
I'm adding a change to the test to enable trace logging in errorlog. It would slowdown that specific test but we'll have logs collected on traces around ipa-pwd-extop.

abbra commented:
For the reference, here is resulting set of tracing logs from executing ``test_commands::TestIPACommand::test_change_sysaccount_password_issue7181` test in Azure Pipelines:
[test_integration-test_commands.py-TestIPACommand-test_change_sysaccount_password_issue7181.tar.gz](https://github.com/freeipa/freeipa/files/4359169/test_integration-test_commands.py-TestIPACommand-test_change_sysaccount_password_issue7181.tar.gz)
I haven't yet got through them.

rcritten commented:
I added a patch to not save password history for non-Kerberos users.

I also added a couple of test cases to ensure that password policy is enforced and that non-kerb users don't get a history.

rcritten commented:
Closing since we can't coax PR-CI into picking this up. Will re-open in another PR.
